lib/instrumentation/modules/aws-sdk/sqs.js

```diff
@@ -0,0 +1,193 @@
+  const toFrom = getToFromFromOperation(request.operation)
+  const queueName = getQueueNameFromRequest(request)
+
+  const name = `${SUBTYPE.toUpperCase()} ${action.toUpperCase()} ${toFrom} ${queueName}`
+  return name
+}
+
+function shouldIgnoreRequest (request, agent) {
+  const operation = request && request.operation
+  // are we interested in this operation/method call?
+  if (OPERATIONS.indexOf(operation) === -1) {
+    return true
+  }
+
+  // is the named queue on our ignore list?
+  if (agent._conf && agent._conf.ignoreMessageQueuesRegExp) {
+    const queueName = getQueueNameFromRequest(request)
+    for (const rule of agent._conf.ignoreMessageQueuesRegExp) {
+      if (rule.test(queueName)) {
+        return true
+      }
+    }
+  }
+
+  if (!agent.currentTransaction) {
+    agent.logger.trace('no active transaction found, skipping sqs instrumentation')
+    return true
+  }
+
+  return false
+}
+
+function wrapReceiveMessageCallback (original, request, agent) {
+  if (typeof original !== 'function' || original.name === 'elasticReceiveMessageCallback') return original
+
+  return elasticReceiveMessageCallback
+  function elasticReceiveMessageCallback (err, data) {
+    recordMetrics(getQueueNameFromRequest(request), data, agent)
+    return original.apply(this, [err, data])
+  }
+}
+
+// Main entrypoint for SQS instrumentation
+//
+// Must call (or one of its function calls must call) the
+// `orig` function/method
+function instrumentationSqs (orig, origArguments, request, AWS, agent, { version, enabled }) {
+  if (shouldIgnoreRequest(request, agent)) {
+    return orig.apply(request, origArguments)
+  }
+
+  const type = TYPE
+  const subtype = SUBTYPE
+  const action = getActionFromRequest(request)
+  const name = getSpanNameFromRequest(request)
+  const span = agent.startSpan(name, type, subtype, action)
+  span.setDestinationContext(getMessagingDestinationContextFromRequest(request))
+
+  const cb = origArguments[0] ? origArguments[0] : null
+  let wrappedCallback = agent._instrumentation.bindFunction(cb)
+
+  // if we're receiving a message we need on more layer of
+  // wrapping in order to access the receiveMessage callback
+  // data and use that data to record metrics
+  if (OPERATIONS_TO_ACTIONS.receiveMessage === action) {
+    wrappedCallback = wrapReceiveMessageCallback(wrappedCallback, request, agent)
+  }
+
+  // replace the argument
+  origArguments[0] = wrappedCallback
+
+  const origResult = orig.apply(request, origArguments)
+
+  // our complete listener MUST be registered after
+  // calling the original Request.send.  Request.send
+  // invokes its callback function by registering its
+  // own complete callback. We need our complete callback
+  // to fire _after_ the one registered by Request.send
+  // see: https://github.com/aws/aws-sdk-js/blob/e5c12fad7ebc87b4de447d777628030be4792945/lib/request.js#L368
+  request.on('complete', function () {
+    span.end()
+  })
+  return origResult
```

Comment:
From our discussion, here is the alternative take that just uses the 'complete' event handler:

- Some "XXX"s to handle with the captureError.
- This (arguably) fixes the span duration to *not* include the user's callback code.

```js
function instrumentationSqs2 (orig, origArguments, request, AWS, agent, { version, enabled }) {
  if (shouldIgnoreRequest(request, agent)) {
    return orig.apply(request, origArguments)
  }

  const type = TYPE
  const subtype = SUBTYPE
  const action = getActionFromRequest(request)
  const name = getSpanNameFromRequest(request)
  const span = agent.startSpan(name, type, subtype, action)
  span.setDestinationContext(getMessagingDestinationContextFromRequest(request))

  request.on('complete', (response) => {
    if (response && response.error) {
      // XXX Have to do the same skipOutcome + // span._setOutcomeFromErrorCapture()
      //     dance as in elasticsearch.js before the "current span" has moved
      //     to the HTTP child span.
      //     (Consider an option to pass the span to `agent.captureError()`)
      // XXX Consider also the same `captureAttributes: false` + explicit custom
      //     fields dance.
      agent.captureError(response.error)
    }

    // We know, from knowledge of the aws-sdk, that this is async.
    span.sync = false

    span.end()

    if (request.operation === 'receiveMessage' && response && response.data) {
      recordMetrics(getQueueNameFromRequest(request), response.data, agent)
    }
  })

  return orig.apply(request, origArguments)
}
```

Reply:
I made all these change _except_ the captureAttributes one.  It feels like a premature optimization and it's uncertain which explicit custom fields would be the right ones. I intend to hold off on that for now

Reply:
+1

I think the logic is backwards and that the agent should always be explicit about what fields to capture on the error. But that's a general and separate issue. Not for here.